cpp/src/indexing/point_quadtree.cu

```diff
@@ -0,0 +1,957 @@
+                                                  cudaStream_t stream)
+{
+  // for each value in `mask` copy from `lhs` if true, else `rhs`
+  auto policy = rmm::exec_policy(stream);
+  auto output = make_fixed_width_column<TypeOut>(size, stream, mr);
+  auto iter   = make_zip_iterator(mask.begin<bool>(), lhs.begin(), rhs.begin());
+
+  thrust::transform(policy->on(stream),
+                    iter,
+                    iter + size,
+                    output->mutable_view().template begin<TypeIn>(),
+                    // return bool ? lhs : rhs
+                    [] __device__(auto const &t) {
+                      return thrust::get<0>(t) ? thrust::get<1>(t) : thrust::get<2>(t);
+                    });
+
+  return output;
+}
+
+inline std::unique_ptr<cudf::experimental::table> make_quad_tree(
+  rmm::device_vector<uint32_t> &quad_keys,
+  rmm::device_vector<uint32_t> &quad_point_count,
+  rmm::device_vector<uint32_t> &quad_child_count,
+  rmm::device_vector<int8_t> &quad_levels,
+  cudf::size_type num_parent_nodes,
+  cudf::size_type const num_levels,
+  cudf::size_type const min_size,
+  cudf::size_type const level_1_size,
+  rmm::mr::device_memory_resource *mr,
+  cudaStream_t stream)
+{
+  auto policy = rmm::exec_policy(stream);
+  // count the number of child nodes
+  auto num_child_nodes = thrust::reduce(
+    policy->on(stream), quad_child_count.begin(), quad_child_count.begin() + num_parent_nodes);
+
+  cudf::size_type num_valid_nodes{0};
+  cudf::size_type num_invalid_parent_nodes{0};
+
+  // prune quadrants with fewer points than required
+  // lines 1, 2, 3, 4, and 5 of algorithm in Fig. 5 in ref.
+  std::tie(num_invalid_parent_nodes, num_valid_nodes) = remove_unqualified_quads(quad_keys,
+                                                                                 quad_point_count,
+                                                                                 quad_child_count,
+                                                                                 quad_levels,
+                                                                                 num_parent_nodes,
+                                                                                 num_child_nodes,
+                                                                                 min_size,
+                                                                                 level_1_size,
+                                                                                 stream);
+
+  num_parent_nodes -= num_invalid_parent_nodes;
+
+  // Construct the indicator output column.
+  // line 6 and 7 of algorithm in Fig. 5 in ref.
+  auto indicator = construct_non_leaf_indicator(
+    quad_point_count, num_parent_nodes, num_valid_nodes, min_size, mr, stream);
+
+  // Construct the f_pos output column
+  // lines 8, 9, and 10 of algorithm in Fig. 5 in ref.
+  auto f_pos = [&]() {
```

Comment:
What does f_pos mean?

Reply:
> What does f_pos mean?
first child position/offset for non-leaf nodes and first point position/offset in the point vertex arrays (x/y) for leaf nodes.

Reply:
I should have phrased this as "use a clearer name, please."